### DISCUSSION TITLE
What is the key/value requirements for these maps?

### DISCUSSION THREAD

**Discussion by afshinpir** [greg7mdp/parallel-hashmap#127] [Q&A] [answered]
Hi

I wonder what is the requirements for the key/value of these maps. For example, can I use them with a structure like this as value?
```cpp
struct S {
    S(int i) : m_(i) {}
    std::atomic<int> m_;
};
```
As you see, this structure is not copy/move constructible/assignable. I can use this with `std::unordered_map`, but I wonder if I can use such structure an my map value.

**Comment by greg7mdp** [comment, marked as answer]
Hi @afshinpir, thanks for the question. Because std::atomic is not movable, you won't be able to use use it in `flat` hash maps (where the values are moved when the map resizes), but you should be able to use them in phmap's `node` hash map, for example in `phmap::node_hash_map`.

**Comment by afshinpir** [comment]
@greg7mdp Thanks for the info.
